### DISCUSSION TITLE
How to use startScreenTrace?

### DISCUSSION THREAD

**Discussion by floydkim** [invertase/react-native-firebase#7470] [Q&A]
Hi everyone,

I need guidance on using the `startScreenTrace` function.
I've checked the [documentation](https://rnfirebase.io/perf/usage#custom-screen-traces), but I'm still unsure about how to apply it to screen components in my app.
Can anyone provide a simple example of how to implement this for tracking screen performance? Any help would be much appreciated!

Thanks!

**Comment by mikehardy** [comment]
I'm sorry I don't understand, you state you have read the documentation, but you want a simple example.

The documentation *contains a simple example*

Perhaps you should describe the tests you have performed (with the code you used), the results you obtained, why they are confusing, and where specifically you need help in order to obtain the results you want related to the tests you provided.

https://stackoverflow.com/help/how-to-ask
https://stackoverflow.com/help/mcve

**Reply by floydkim** [reply]
@mikehardy 

I'm sorry for asking the question without providing enough context.

After reviewing the docs, I'm still unsure about the correct way to use `startTrace` function in React components.

<details>
  <summary>the example in docs</summary>

```ts
async function screenTrace() {
  // Define & start a screen trace
  try {
    const trace = await perf().startScreenTrace('FooScreen');
    // Stop the trace
    await trace.stop();
  } catch (e) {
    // rejects if iOS or (Android == 8 || Android == 8.1)
    // or if hardware acceleration is off
  }
}
```

</details>

Here's how I attempted to use it:
```tsx
function FooScreen() {
    screenTrace()

    return <SomeView />
}
```

However, I'm concerned that calling start and stop in rapid succession, as shown in the documentation's example, may not provide accurate measurements. I was expecting an implementation that aligns with the react component's mount-unmount lifecycle.

Also I'm unsure whether this function should be called within the React component or outside it.

Could you clarify the best practice for using `startTrace` in a React component to track screen performance accurately?

**Reply by anujmpec** [reply]
@floydkim Did you get the proper result  from performance monitoring?

**Reply by floydkim** [reply]
@anujmpec no 🥲 It's out of my interest now.

**Reply by fabioatcorreia** [reply]
As stated [here](https://github.com/invertase/react-native-firebase/issues/3371#issuecomment-2210123386), it would really be helpful to have a guide on how to setup screen traces for some of the most used react native navigation libraries, since currently it is not very clear.    
As an example, [Datadog docs](https://docs.datadoghq.com/real_user_monitoring/application_monitoring/react_native/advanced_configuration/#track-view-navigation) follow this kind of approach.

**Comment by stee1ix** [comment]
I have a custom implementation. I use onStateChange of NavigationContainer and start on a route mount and stop on route change. I'm not sure if this works correctly or not. I inherited this piece of code. Does the tracer continues with the instance till stop is called, or we need to let the await on startScreenTrace handle itself? @mikehardy 
```javascript
// screenTrace.js
import perf from '@react-native-firebase/perf';
import crashlytics from '@react-native-firebase/crashlytics';

async function screenTrace(screenName) {
  try {
    const trace = await perf().startScreenTrace(screenName);
    return {
      async stop() {
        try {
          await trace.stop();
        } catch (err) {
          crashlytics().recordError(err)
        }
      },
    };
  } catch (err) {
    crashlytics().recordError(err)
    return null;
  }
}

export default screenTrace;


// onStateChange
if (previousRouteKey !== currentRouteKey) {
      if (!prevRoutePerfRef.current) {
        screenTrace(currentRouteName).then((trace) => {
          prevRoutePerfRef.current = trace;
        });
      } else {
        // Close the previous trace
        prevRoutePerfRef.current?.stop?.().then(() =>
          screenTrace(currentRouteName).then((trace) => {
            prevRoutePerfRef.current = trace;
          })
        );
      }
}
